src/main/java/org/gridsuite/timeseries/server/TimeseriesDataRepository.java

```diff
@@ -0,0 +1,347 @@
+        } else {
+            // TODO here if we try with 51 cols we get an exception
+            // Caused by: org.postgresql.util.PSQLException: ERROR: cannot pass more than 100 arguments to a function
+            return col.stream().map(c -> "'" + c + "', json_obj->'" + c + "'").collect(Collectors.joining(", ",
+                    "select time, json_build_object (",
+                    ") json_obj from timeseries_group_data where group_id=? and time>=? and time <? order by time;"));
+        }
+    }
+
+    private final ObjectMapper objectMapper;
+    private final HikariDataSource datasource;
+    private final TimeSeriesMetadataService timeseriesMetadataService;
+
+    public TimeseriesDataRepository(ObjectMapper objectMapper, HikariDataSource datasource,
+            TimeSeriesMetadataService timeseriesMetadataService) {
+        this.objectMapper = objectMapper;
+        this.datasource = datasource;
+        this.timeseriesMetadataService = timeseriesMetadataService;
+    }
+
+    // TODO tune these parameters for performance
+    // TODO make these parameters in application.yaml
+    private static final int WRITE_BATCHSIZE = 30000;
+    private static final int WRITE_THREADSIZE = 3; // 3 batches => e.g. 300 rows of 300cols
+    private static final int READ_THREADSIZE = 300; // 300 db rows, TODO take the number of cols into account
+
+    public void save(UUID uuid, List<TimeSeries> listTimeseries) {
+        try {
+            dosave(uuid, listTimeseries);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // TODO untangle multithreaded scatter/gather from actual work
+    private void dosave(UUID uuid, List<TimeSeries> listTimeseries) throws Exception {
+
+        int colcount = listTimeseries.size();
+        int rowcount = listTimeseries.get(0).getMetadata().getIndex().getPointCount();
+
+        int batchrow = (WRITE_BATCHSIZE + colcount - 1) / colcount;
+        int batchcount = (rowcount + batchrow - 1) / batchrow;
+
+        int threadcount = (batchcount + WRITE_THREADSIZE - 1) / WRITE_THREADSIZE;
+        int threadbatches = (batchcount + threadcount - 1) / threadcount;
+
+        List<Callable<Void>> callables = new ArrayList<>(Collections.nCopies(threadcount, null));
+
+        LOGGER.debug(
+                "insert {} in batch of {} rows ({} doubles for each batch), numbatch={}, numthreads={}, threadbatches={}",
+                uuid, batchrow, batchrow * colcount, batchcount, threadcount, threadbatches);
+        long a = System.nanoTime();
+
+        // TODO here we transpose, which means it's impossible to stream
+        // data from the client to the database, the server has to buffer in memory.
+        // try to change the API to allow streaming.
+        // TODO avoid copying the data (timeseries toArray())?
+        BiFunction<Integer, Integer, Object> stringOrDoubledataGetter;
+        if (listTimeseries.get(0) instanceof DoubleTimeSeries) {
```

Comment:
in metadata, we have info of dataType, is possible avoid check with the first element by get(0)
"metadata" : {
      "name" : "_GEN____1_SM_generator_UStatorPu",
      "dataType" : "DOUBLE",
      "tags" : [ ],
      "irregularIndex" : []
    }